reactor-core/src/main/java/reactor/core/Disposable.java

```diff
@@ -165,4 +167,11 @@ default boolean addAll(Collection<? extends Disposable> ds) {
 		 */
 		int size();
 	}
+
+	// TODO(dj): add javadoc
```

Comment:
the semantics of disposeGracefully may vary widely, so I would make the documented contract say that explicitly (eg. "each class implementing this trait should define how subsequent calls behave during the grace period and after it")

Reply:
one limitation I'm thinking of with this API is that most of the time the underlying resource(s) being disposed gracefully will be atomically swapped out. Which means that even if one re-subscribes to the `Mono`, including with `onErrorResume()` or `retry()`, the underlying resources won't be reachable anymore.

thus, there will be no way of composing operators to fall back to a "hard" shutdown once the graceful shutdown is initiated.

I'm thinking this is fine if documented. The recommendation for implementors should probably be to trigger a hard dispose at the end of the `gracePeriod` THEN propagate a `TimeoutException`, noting that it only serves as a warning / logging but cannot be recovered.

Reply:
We discussed the possible contracts here with @OlegDokuka. The approach with forceful shutdown before propagating the `TimeoutException` and non-recoverable errors might be confusing to users if they don't read the specific `Scheduler` documentation, but it has some advantages implementation wise.
Another approach could be to propagate a retry-able error and allow re-initiating the shutdown + await procedure, while also allowing for an explicit final call to explicit `shutdownNow()` when desired. I'll go back to the original issue and ask for opinion from the user's perspective to guide the design.

Reply:
My personal opinion is that once we fix specific behavior (e.g. call `shutdownNow()` if timeout or interaction exception) then we probably will end up with everyone doing `scheduler.disposeGracefully(Duration.ofHours(9999999)).subscribe()` or then complaining that they did not wont to have `shutdnowNow` called but rather retry later 

Another thought on `TimoutException` - any exception is useless if we can not do anything useful after that. I'm not sure that logging such an event makes any sense. This exception is just a fact that we forced shutdown process so use just has to take it. Also, taking into account the impl details - all other active subscribers are going to get the same notification but the other late subscriber will not get it, then it is going to be too confusing so even having it documented will not resolve this confusion.

My personal recommendation is to prefer flexibility over fixed behavior. One can always write something like the following to mimic what we can hardcode

```
scheduler.disposeGracefully(Duration.ofMillis(100))
    .retryWhen(Retry.backoff(5, Duration.ofMillis(50)))
    .onErrorResume(e -> Mono.fromRunnable(scheduler::dispose));
```